571 variable interpolation

iandjx opened the pull request:
- Enhanced the CrewAppForm component to handle variable choices and kickoff variables, including UI updates.
  - Integrated variable selection and creation in SortableItem and TaskForm components.
  - Added session variable update logic in StructuredInputForm with API integration.
  - Refactored task building logic in agent-backend with helper functions and session variable interpolation.
  - Simplified variable fetching logic in the database layer.
  - Added new helper functions for task management in agent-backend.
  - Improved session and task API logic to handle new variable-related features.

tomlynchRNA commented on webapp/src/controllers/task.ts:
```diff
@@ -281,9 +287,10 @@ export async function addTaskApi(req, res, next) {
+		formFields: formFieldsWithObjectIdVariables,
 		isStructuredOutput,
-		variableIds: (variableIds || []).map(toObjectId)
+		variableIds: (variableIds || []).map(toObjectId),
```
A variable ID here can refer to a variable that was deleted

tomlynchRNA commented on webapp/src/components/CrewAppForm.tsx:
```diff
@@ -115,6 +122,33 @@ export default function CrewAppForm({
+	const selectedAgents = agentIds?.map(a => agentChoices.find(ac => ac._id === a));
+	const variableIdsOfSelectedAgents = selectedAgents?.map(a => a?.variableIds || []).flat();
+	const variablesOfSelectedAgents = variableIdsOfSelectedAgents?.map(v =>
+		variableChoices?.find(vc => vc._id === v)
```
The variable may not be found, suggest to filter this list for safety but also handle this in backend when saving tasks, agents, etc

tomlynchRNA commented on webapp/src/components/CrewAppForm.tsx:
```diff
@@ -485,6 +520,53 @@ export default function CrewAppForm({
+												setKickOffVariables(v || []);
+											}}
+											options={combinedVariables.map(v => ({
+												label: v.name,
```
Uncaught error here because the variable can be null see above comment

tomlynchRNA commented on webapp/src/controllers/session.ts:
```diff
@@ -86,7 +87,12 @@ export async function sessionData(req, res, _next) {
-				crewAppVariables.push(...variables);
+
+				variables.forEach(variable => {
+					if (!crewAppVariables.some(v => (v._id as ObjectId).equals(variable._id))) {
```
variable can be null, need to handle or filter above

tomlynchRNA reviewed (changes requested):
comments apply to agents and tasks even if I only commented one one

iandjx commented:
> comments apply to agents and tasks even if I only commented one one

Added changes such that error gets returned if a non-existent variable is included when adding  a new agent/task or editing an agent/task

Added logic to fitler out undefined variables in crewapp form

tested by 
1. deleting a variable before hitting save when adding/editing a task/agent
2. deleting a variable before creating a new app
